### DISCUSSION TITLE
Add an endpoint to get a database dump for backup purposes

### DISCUSSION THREAD

**Discussion by vs49688** [navidrome/navidrome#2258] [Ideas]
As in title. It would be extremely useful to add an endpoint (such as `/admin/backup`) that returns a snapshot-in-time dump of the database. This would enable consistent backups without having to shutdown/restart Navidrome, e.g. I could hit the endpoint in a daily cron job.

SQLite has the [Backup API](https://www.sqlite.org/backup.html). A GET to the endpoint could simply return a binary blob.
Everything under `/admin` would obviously require the user to be an admin.

I just took a crack at this, however I'm not familiar with beego, and I'm not entirely sure how (and even if) it can use SQLite's backup API.

(Alternatively, if Navidrome supported Postgres this would be moot, as I could do it as part of my existing database backup ;))

**Comment by deluan** [comment]
Not sure if this is a good thing, in terms of security. And I'm not 100% sure this is really needed based on Navidrome's use of SQLite3.

I've successfully backed up my DB by simply zipping all `navidrome.db*` files without closing Navidrome. Every time I need it, the backup has been restored without any issues. The secret is to ensure that the WAL and SHM files are not forgotten.

Just to give you a proof that this is indeed secure, until version 0.48.0 Navidrome didn't close the DB when stopped, basically shutting down the SQLite3 engine without merging the DB and WAL files. And I *never* received any corrupted data report because of this.

**Reply by vs49688** [reply]
> Not sure if this is a good thing, in terms of security.

Oh it's a risk, absolutely. If it were done, there should probably a separate `/admin` endpoint that is listening on another port and not exposed to the internet.

And I'm not 100% sure this is really needed based on Navidrome's use of SQLite3.

Completely up-to-you, it was just a suggestion.

> I've successfully backed up my DB by simply zipping all `navidrome.db*` files without closing Navidrome. Every time I need it, the backup has been restored without any issues. The secret is to ensure that the WAL and SHM files are not forgotten.
> 
> Just to give you a proof that this is indeed secure, until version 0.48.0 Navidrome didn't close the DB when stopped, basically shutting down the SQLite3 engine without merging the DB and WAL files. And I _never_ received any corrupted data report because of this.

I've done backups by copying the files too, and it's always worked. It just... seems wrong. Normally, on another RDBMS like Postgres, I'd just run `pg_dumpall` and get snapshot-in-time SQL dump I can restore, but of course Navidrome only supports SQLite3.

Here's my actual motive behind this:

I have periodic backups running using `systemd`. For security purposes, I run these with `PrivateUsers=true`. For backup up Postgres, I `pg_dumpall` into a temporary directory (with `PrivateTmp=true`) and upload it to S3. I've been looking to do the same with Navidrome, and am _really_ not comfortable giving anything access to the filesystem that doesn't strictly need it. Hence a way of dumping the DB without FS access.

I can work around this, of course, but I thought I'd suggest a potential solution. I'm probably too paranoid, so feel free to ignore this :)

**Reply by dani** [reply]
Having either a dump API endpoint (or even better: support postgres!) would also makes things simpler when running in a container orchestrator. As sqlite do not play well with nfs or other shared fs, we need to use a block based storage, which means the FS cannot be accessed from another pod/allocation (even read only). I'm having this problem with Nomad, but it'd be the same with k8s. For now, the only solution is to add a small cron based backup container in the same pod as navidrome.

**Comment by metalheim** [comment]
Backup function exists now. https://www.navidrome.org/docs/usage/backup/
We should close this :)
